poh/src/poh_recorder.rs

```diff
@@ -576,6 +576,15 @@ impl PohRecorder {
             return PohLeaderStatus::NotReached;
         }
 
+        if self.blockstore.has_existing_shreds_for_slot(next_poh_slot) {
+            // We already have existing shreds for this slot. This can happen when this block was previously
+            // created and added to BankForks, however a recent PoH reset caused this bank to be removed
+            // as it was not part of the rooted fork. If this slot is not the first slot for this leader,
+            // and the first slot was previously ticked over, the check in `leader_schedule_cache::next_leader_slot`
+            // will not suffice, as it only checks if there are shreds for the first slot.
```

Comment:
Interestingly enough `fn next_leader_slot` returns (start_slot, last_slot) and `last_slot` is used to calculate `leader_last_tick_height` in `PohRecorder` but we only use `leader_last_tick_height` inside `would_be_leader`. Maybe `reached_leader_tick` needs to use it as well?

Reply:
It could be an option, it would allow us to remove the check here for when we tick into the next leader:
https://github.com/anza-xyz/agave/blob/c99095da982e23b4f2df0d8f4dfa19032b316659/core/src/replay_stage.rs#L2073-L2075

However I think if we use it in `reached_leader_tick`, we should recompute our next leader window. Otherwise if we don't reset in time, we could end up missing our next leader window, which the current code prevents.